plugins/synaptics-cape/fu-synaptics-cape-firmware.c

```diff
@@ -0,0 +1,215 @@
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VID,
+					&self->vid,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_PID,
+					&self->pid,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_X,
+					&self->version_x,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_Y,
+					&self->version_y,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_Z,
+					&self->version_z,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	self->version = g_strdup_printf("%u.%u.%u.%u",
+					self->version_z,
+					self->version_y,
+					self->version_x,
+					self->version_w);
+	fu_firmware_set_version(FU_FIRMWARE(self), self->version);
+
+	fw_hdr = fu_common_bytes_new_offset(fw, 0, sizeof(FuCapeHidFwCmdUpdateStartPar), error);
+	if (fw_hdr == NULL)
+		return FALSE;
+
+	fu_firmware_set_id(img_hdr, FU_FIRMWARE_ID_HEADER);
+	fu_firmware_set_bytes(img_hdr, fw_hdr);
+	fu_firmware_add_image(firmware, img_hdr);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_cape_firmware_parse(FuFirmware *firmware,
+				 GBytes *fw,
+				 guint64 addr_start,
+				 guint64 addr_end,
+				 FwupdInstallFlags flags,
+				 GError **error)
+{
+	FuSynapticsCapeFirmware *self = FU_SYNAPTICS_CAPE_FIRMWARE(firmware);
+	const gsize bufsz = g_bytes_get_size(fw);
+	g_autoptr(GBytes) fw_header = NULL;
+	g_autoptr(GBytes) fw_body = NULL;
+	const gsize headsz = sizeof(FuCapeHidFileHeader);
+
+	/* check minimum size */
+	if (bufsz < sizeof(FuCapeHidFileHeader)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "not enough data to parse header, size ");
+		return FALSE;
+	}
+
+	if ((guint32)bufsz % 4 != 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "data not aligned to 32 bits");
+		return FALSE;
+	}
+
+	fw_header = g_bytes_new_from_bytes(fw, 0x0, headsz);
+
+	if (fw_header == NULL) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE, "out of memory");
+		return FALSE;
+	}
+
+	if (!fu_synaptics_cape_firmware_parse_header(self, firmware, fw_header, error))
+		return FALSE;
+
+	fw_body = g_bytes_new_from_bytes(fw, headsz, bufsz - headsz);
+
+	fu_firmware_set_id(firmware, FU_FIRMWARE_ID_PAYLOAD);
+	fu_firmware_set_bytes(firmware, fw_body);
+	return TRUE;
+}
+
+static void
+fu_synaptics_cape_firmware_init(FuSynapticsCapeFirmware *self)
+{
+	fu_firmware_add_flag(FU_FIRMWARE(self), FU_FIRMWARE_FLAG_HAS_VID_PID);
+}
+
+static void
+fu_synaptics_cape_firmware_finalize(GObject *object)
+{
+	FuSynapticsCapeFirmware *self = FU_SYNAPTICS_CAPE_FIRMWARE(object);
+
+	g_free(self->version);
```

Comment:
I don't see a good reason to have `self->version`.  Just use `fu_firmware_set_version` and leave it at that

Reply:
okay, will use ```fu_firmware_set_version``` instead.